org.lflang.tests/src/org/lflang/tests/TestBase.java

```diff
@@ -467,19 +468,28 @@ protected void addExtraLfcArgs(Properties args) {
 
     /**
      * Invoke the code generator for the given test.
+     *
      * @param test The test to generate code for.
      */
-    private GeneratorResult generateCode(LFTest test) {
-        GeneratorResult result = GeneratorResult.NOTHING;
-        if (test.fileConfig.resource != null) {
+    private GeneratorResult generateCode(LFTest test) throws TestExecutionException {
+        if (test.fileConfig.resource == null) {
+            return GeneratorResult.NOTHING;
+        }
+
+        try {
             generator.doGenerate(test.fileConfig.resource, fileAccess, test.context);
-            result = test.context.getResult();
-            if (generator.errorsOccurred()) {
-                test.result = Result.CODE_GEN_FAIL;
-                throw new AssertionError("Code generation unsuccessful.");
-            }
+        } catch (AssertionError | Exception e) {
+            test.result = Result.CODE_GEN_FAIL;
+            // Add the stack trace to the test output
+            e.printStackTrace(new PrintWriter(test.compilationLog, true));
```

Comment:
It would be cleaner to add a method `handleCodeGenException(Throwable)` to LFTest that does this, instead of making the compilation log public.

Also this logic would probably better be moved to the parent method, and here a TestExecutionException be thrown that has the exception `e` as its cause